Allow folding in search

grahnen opened the pull request:
I use a custom search to manage my todos. Folding the items from the search panel (using the same logic for folding in notes) makes the search screen more usable for me. Added an option to turn it on/off, which defaults to off.

ShuoShen commented on app/src/main/java/com/orgzly/android/prefs/AppPreferences.java:
```diff
@@ -854,6 +854,21 @@ public static void isNoteContentFolded(Context context, boolean value) {
+
+
+    public static boolean isSearchFoldable(Context context) {
+        return getStateSharedPreferences(context).getBoolean(
```
It's probably more straightforward to use `getDefaultSharedPreferences` for values stored in the setting menu (`isNotesContentFoldable` as an example: https://github.com/orgzly-revived/orgzly-android-revived/blob/f2d7d1767633d673726f0a4b896e472d9813b02c/app/src/main/java/com/orgzly/android/prefs/AppPreferences.java#L179-L183). 

On the other hand, `getStateSharedPreferences` is intended for storing app states, which is suited for things like `isFolded` (`isNoteContentFolded` as an example: https://github.com/orgzly-revived/orgzly-android-revived/blob/f2d7d1767633d673726f0a4b896e472d9813b02c/app/src/main/java/com/orgzly/android/prefs/AppPreferences.java#L846-L855)

ShuoShen replied:
Ah, I just realized that the next commit updated this to getDefaultSharedPreferences. Nice. 

Maybe squash the commits into one to keep the history from getting too cluttered?

ShuoShen commented on app/src/main/java/com/orgzly/android/ui/notes/NoteItemViewBinder.kt:
```diff
@@ -314,7 +314,7 @@ class NoteItemViewBinder(private val context: Context, private val inBook: Boole
         var isVisible = false
 
-        if (inBook) {
+        if (AppPreferences.isSearchFoldable(context) || inBook) {
```
I might be misunderstanding the purpose of the diff. Is it for hiding the search query names on the left-hand-side panel? This code could change the behavior of NoteItem, which doesn't match my understanding. 

BTW, for my own undrestanding, do you mind sharing a few screenshots to illustrate the visual changes?

grahnen replied:
This line is for drawing the folding-buttons in the search, too.

ShuoShen commented on app/src/main/java/com/orgzly/android/prefs/AppPreferences.java:
```diff
@@ -854,6 +854,21 @@ public static void isNoteContentFolded(Context context, boolean value) {
+
+    public static void isSearchFoldable(Context context, boolean value) {
+        String key = context.getResources().getString(R.string.pref_key_is_search_foldable);
+        getStateSharedPreferences(context).edit().putBoolean(key, value).apply();
```
This can be removed? It doesn't appear to be used anywhere.

ShuoShen reviewed (changes requested):
Thanks a lot for contributing! The description makes sense to me, and thanks for putting the feature behind a new setting. 

However, I might have a slight misunderstanding of what experience this PR will enable. Do you mind attaching a few screenshots to illustrate the change? 

It would be nice if @amberin can take a look at this, who should have more context.

grahnen commented:
> Thanks a lot for contributing! The description makes sense to me, and thanks for putting the feature behind a new setting.
> 
> However, I might have a slight misunderstanding of what experience this PR will enable. Do you mind attaching a few screenshots to illustrate the change?
> 
> It would be nice if @amberin can take a look at this, who should have more context.

![Unfolded](https://github.com/user-attachments/assets/b4bdee85-a76a-4884-bf5f-7775a994a7ab)
![Folded](https://github.com/user-attachments/assets/2ebcbfa5-fa8b-4658-afda-85aefe31a42d)

Clicking the folding button (-/+) folds and unfolds the content of the note, respectively.

It uses the folded state of the note itself, meaning if you fold it in the search it also folds in the corresponding notebook.

I squashed the commits and removed the unused function

amberin commented:
Hello @grahnen, and welcome to the Orgzly community! I didn't have time to build this myself, but from your screenshots it looks really nice to me!

I'm not sure if anyone could see issues with this, but I suggest we release the feature and then simply await feedback.

Big thanks for your contribution.

ShuoShen commented on app/src/main/java/com/orgzly/android/prefs/AppPreferences.java:
```diff
@@ -854,6 +854,17 @@ public static void isNoteContentFolded(Context context, boolean value) {
+    /*
+     * Folding in search
+     */
+
+
```
nit: delete additional whitespaces.

grahnen replied:
For the other such comments, there is one whitespace below and one after, so I'll format this the same way.

ShuoShen commented on app/src/main/java/com/orgzly/android/ui/notes/NoteItemViewBinder.kt:
```diff
@@ -314,7 +314,7 @@ class NoteItemViewBinder(private val context: Context, private val inBook: Boole
         var isVisible = false
 
-        if (inBook) {
+        if (AppPreferences.isSearchFoldable(context) || inBook) {
```
> AppPreferences.isSearchFoldable(context) || inBook

Seeing this check in a few different places makes me wonder if it's a good tradeoff to add a new app preference for `isSearchFoldable`. My gut feeling is we can probably just implement this feature without introducing a new pref. 

In order to trigger this feature, we need to turn on two other preferences:  'Display content in Search', 'Fold content'. And now 'is search foldable'. This feature has minimal UX impact (i.e. it only adds a foldable button to people who don't want to use it), and it's already heavily gated by preferences. I feel like it's not worth a new preference. Adding a new preference will clutter the setting page, making it more difficult to read the code and reason about user behaviors. 

What do you think @amberin?

grahnen replied:
Since the fold status is inherited from the status of the note itself, folding it in a notebook will have it show up as folded in searches. This might not be something that's wanted as default behaviour. That was my rationale for adding it as a preference. Alternatively, one can add a new property to NotePosition (isFoldedInSearch or similar name), to have these two be separate. Which solution we go for is up to you.

ShuoShen replied:
I see. That makes sense. Thanks for putting thoughts to it. 

Another preference to consider is whether we should enable folding in search but allowing users to default to search results folded.  There's already a "Start folded" preference, which is currently dedicated to books only. There are a few options to do that: 
1. Allow "Start folded" to control the search behavior as well
2. Change the description of "Start folded" to clearly indicate it's for notes only, so that the search users will think it's a bug that they turn this on but search results are all expanded
3. If we go with 2, we might want to add another option for "Search start folded"

I would say maybe this could be a follow up PR. This PR looks good to me.

ShuoShen commented:
@amberin Likewise. Great to have you @grahnen!

The code looks good to me. Main question is whether we want to add another preference. Looks like it introduces more problems than it solves (please see comments on the commit). Unless we have concern this is going to break something, I'd say maybe just roll it out without the new feature, or have a plan to remove feature after successful launch.

grahnen commented on app/src/main/res/xml/prefs_screen_notebooks.xml:
```diff
@@ -82,6 +82,13 @@
             android:summary="@string/display_both_title_and_content"
             android:defaultValue="@bool/pref_default_is_notes_content_displayed_in_list"/>
 
+        <SwitchPreference
+            android:key="@string/pref_key_is_notes_content_foldable"
+            android:dependency="@string/pref_key_is_notes_content_displayed_in_list"
+            android:title="@string/fold_content"
+            android:summary="@string/allow_folding_of_content"
+            android:defaultValue="@bool/pref_default_is_notes_content_foldable"/>
+
         <SwitchPreference
```
The rationale was to keep the `is_notes_content_foldable` preference adjacent to its dependency.

grahnen commented:
Thanks for the warm welcome!

I rebased against latest master. Trying to keep this up to date until it gets merged. (Should I?)

ShuoShen commented:
> Thanks for the warm welcome!
> 
> I rebased against latest master. Trying to keep this up to date until it gets merged. (Should I?)

This has been pushed to the master branch of *your fork*, which is fine. 

But it won't be released just yet. The normal workflow is for this PR to be approved by @amberin, and then he will merge your PR into the orgzly-android-revived upstream repo, manually cut a release branch and launch to the playstore (or fdroid) using the release build. 

I believe @amberin is on a tight schedule now, and we will just wait for him to take over the release process. 

As I mentioned in another thread, I'm happy help scaling the release process if you would like to @amberin. :)

grahnen commented:
> > Thanks for the warm welcome!
> > I rebased against latest master. Trying to keep this up to date until it gets merged. (Should I?)
> 
> This has been pushed to the master branch of _your fork_, which is fine.

Yeah, I meant I'd keep my fork up-to-date with upstream master, and asking if that's necessary. 

> 
> But it won't be released just yet. The normal workflow is for this PR to be approved by @amberin, and then he will merge your PR into the orgzly-android-revived upstream repo, manually cut a release branch and launch to the playstore (or fdroid) using the release build.
> 
> I believe @amberin is on a tight schedule now, and we will just wait for him to take over the release process.

Thanks for letting me know! I did not intend to rush or push you,  sorry for making it seem as such!

>
> As I mentioned in another thread, I'm happy help scaling the release process if you would like to @amberin. :)

amberin reviewed (approved):
Looks excellent, and since it's opt-in, it's practically zero risk. Merging now. Big thanks @grahnen!
